types/express-serve-static-core/index.d.ts

```diff
@@ -181,12 +189,7 @@ export interface RequestRanges extends RangeParserRanges { }
 
 export type Errback = (err: Error) => void;
 
-export interface Dictionary<T> { [key: string]: T; }
-export type ParamsDictionary = Dictionary<string>;
-export type ParamsArray = string[];
-export type Params = ParamsDictionary | ParamsArray;
-
-export interface Request extends http.IncomingMessage, Express.Request {
+export interface Request<P extends Params = ParamsDictionary> extends http.IncomingMessage, Express.Request {
```

Comment:
Oh, one thing I was going to suggest - when this broke for me, I ctrl-clicked in VSCode to get to the type defition which brought me here.  It might be worthwhile putting a comment in here (or maybe above `interface Request`) giving a brief description of how to use this.  Something along the lines of:

```
/**
 * An Express Request.
 * @param {P} For most requests, this should be `ParamsDictionary`, but if you're
 * using this in a route handler for a route that uses a RegExp or a wildcard path
 * (e.g. "/user/*"), then `req.params` will be an array, in which case you should
 * use `Request<ParamsArray>` instead.
 *
 * @example
 *     app.get(/echo/(.*),
 *         (req: Request<ParamsDictionary>, res: Response) => res.send(req.params[1])
 *     );
 */
```

Reply:
The example snippet has an invalid regex (guessing you intended `/echo\/(.*)/`?) and should use `ParamsArray`, not `ParamsDictionary`

If you add this documentation, I suggest adding another snippet for the more common case, to demonstrate that `ParamsDictionary` does not need to be specified explicitly.

Reply:
Good idea. That's now done, let me know what you think.

I put the comment on `Request`, although some users might look to `app.get` (and its other methods), so I'm not sure what the best place for this is.